### PR TITLE
Preventing unnecessary differentials

### PR DESCRIPTION
**Pull request by ngfreiter**
This PR solves the issue described here https://standardhealthrecord.atlassian.net/projects/CIMPL/issues/CIMPL-81?filter=allissues&orderby=priority%20DESC and here https://chat.fhir.org/#narrow/stream/197290-cimpl/topic/CIMPL.206.2E8.2E0/near/176564149. In cases where the `type` was set in the CIMPL to be equal to the same choice of `type`s that was originally on the base structure, `type` was added to the differential, even though the `type` was not modified from its original value. This fixes that bug.

To test, do one build of mCODE using master, and one build using this branch (no need to generate the IG, just build). Then compare those output directories with `diff -r out-1/ out-2/`. You should see that the data dictionaries differ. You should also differences in `DiagnosticReport`, `CBCWAutoDifferentialPanel`, `ComprehensiveMetabolic2000SerumOrPlasmaPanel`, `LaboratoryReport`, and `GenomicsReport`. This difference should be that on the `differential`, on the `performer` field, `type` is removed. You can convince yourself that this is actually correct, because `DiagnosticReport` does not change the `type` of `performer` from the profile it is based on defined here: http://hl7.org/fhir/us/core/STU3/StructureDefinition-us-core-diagnosticreport-note.html, so `type` should not have been showing up on `performer` in the differential, and now it correctly does not appear. All the other classes inherit from `DiagnosticReport`, so they see this change as well. You should see no other differences.

I am not in love with how this fix is done, deleting `type` after the fact. However, `type`s are added to the differential recursively one at a time, so I don't see how we can prevent `type`s from being added, since we don't know until after all `type`s are added if we are going to match the original definition.

### CODE REVIEW

**Review by cmoesel — CHANGES_REQUESTED** [review]
This seems insufficient to me, unless I'm missing something.

My primary concern is that `targetProfilesSameAsOriginal` only looks at the _first_ element in the type array (e.g., `ssType[0]`).  Don't we need to ensure the whole array is the same?

My second concern is that `targetProfile` is not an array in all versions of FHIR, so I think you need to use `MVH.typeTargetProfile(ssType)` to access it in order to guarantee you always get an array value.

My last concern is that we're not looking at the type's `profile` property.  The `targetProfile` is usually only used for a reference, where as `profile` is used for inline profiles.  If we have profiles, and they all match, then that might also be a place to remove the type.

**Comment by ngfreiter** [comment]
The new commit, https://github.com/standardhealth/shr-tools/pull/4/commits/66afa987c6b6229320347f92caed636da193470c, is basically a complete refactoring of the function.

It does now loop over every type and look at all of them.

`MVH` is used for the `targetProfile` and `profile`.

We also look at `profile`.

These changes do not actually change the expected behavior that is described above, still look to `DiagnosticReport` and its children to see the fix in action. I also rebased onto master.

File: packages/shr-fhir-export/lib/export.js [hunk 1]
```diff
@@ -4591,6 +4595,41 @@ function isTargetBasedOn(target, baseTarget, targetSpec, configuration) {
...
+      // Alphabetically sort to make comparison easy
+      const sortedOriginalProfiles = [...originalProfiles].sort();
+      const sortedCurrentProfiles = [...profiles].sort();
+      // See if the arrays are the same
+      targetsMatch = sortedOriginalProfiles.every((el, i) => el === sortedCurrentProfiles[i]);
+    } else if (originalProfiles == null && profiles == null) {
+      profilesMatch = true;
+    }
+    // If allTypesMatch is ever set to false, we want it to stay false
+    allTypesMatch = allTypesMatch && profilesMatch && targetsMatch;
```
Comment (cmoesel) [inline packages/shr-fhir-export/lib/export.js]: Wouldn't it be more efficient just to immediately return the first time you get a `false`?  E.g.,
```
if (!targetsMatch || !profilesMatch) {
  return false;
}
```
Then you skip all remaining iterations.  And you don't need to track an `allTypesMatch` variable.

Given that this function will get called a _lot_, I think it makes sense to go ahead and make this optimization.

**Review by cmoesel — CHANGES_REQUESTED** [review]
I'd like one small optimization.  And can you rebase again?

File: packages/shr-fhir-export/lib/export.js (hunk 1 above)
Comment (ngfreiter) [reply packages/shr-fhir-export/lib/export.js]: Huh, what was I thinking when I did this the first time? Thanks for catching that. Fixed and rebased.

**Review by cmoesel — APPROVED** [review]
Looks good!  I noticed that we couldn't reproduce the _exact_ issue that Kurt reported (dealing with `basedOn`), but this does seem to address the issue in general.

**Pull request merged**